### PR TITLE
Added treatment of bids-incomatible files in dataset

### PR DESCRIPTION
**Pull request by nbeliy**
This pull request addresses issue #267 

Main changes:

- Removed `verbose` option from `parce_filename`, now warnings are always shown
- Slightly changed entities treatment:
  1. removing extention
  2. splitting basename into entities parts
  3. splitting each part into key-value
  4. based on number of keys-values pairs determining if it's suffix or entity
- based on above, providing more explicit warning message
- After `parse_filename` is called, checking if returned structure is empty
- Adding `'tolerant', false` where it make sense

Standing issues:

- if malformed files are json ones, they are rechecked regularly, spamming terminal output

### CODE REVIEW

**Comment by Remi-Gau** [comment]
General question:

What SHOULD be the expected behavior when of `parse_filename` when using it in `tolerant` mode on a bids invalid file?

Would it make sense for `parse_filename` to at least return `p.filename = filename` in some casess?

The reason I am wondering about it is also to because what should be the expected behavior of the following?

```
tolerant = true;
use_schema = false;
index_derivatives = false;
BIDS = layout(path_to_bids, use_schema, index_derivatives, tolerant);

data = bids.query(BIDS, 'data')
```

My assumption is that we would want this to list all the files that are in the dataset no matter if they are bids valid or not?

**Comment by Remi-Gau** [comment]
Will try fixing the tests first and add some more to cover the blind spots that this PR adresses

**Comment by nbeliy** [comment]
> What SHOULD be the expected behavior when of `parse_filename` when using it in `tolerant` mode on a bids invalid file?

I like current state: tolerant switches from error to warning, and the function calling `parse_filename` decide how to act on tolerated files. My initial goal is just report the files that failed instead of just random non descript error. For strict bids datasets it should produce errors, but for non strict, these files can be just ignored (but still reported) 

> Would it make sense for `parse_filename` to at least return `p.filename = filename` in some casess?

Not sure, the `[]` clearly indicates failed state, if the return will contain partially filled structure, the test for failed state will be more intricated. the function that calls `parse_filename` knows already the file name to parse

> The reason I am wondering about it is also to because what should be the expected behavior of the following?

I'll say it should put in dataset maximum files that still follow bids syntax. Just makes it easier for unregulated bids datasets.

**Comment by Remi-Gau** [comment]
> > What SHOULD be the expected behavior when of `parse_filename` when using it in `tolerant` mode on a bids invalid file?
> 
> I like current state: tolerant switches from error to warning, and the function calling `parse_filename` decide how to act on tolerated files. My initial goal is just report the files that failed instead of just random non descript error. For strict bids datasets it should produce errors, but for non strict, these files can be just ignored (but still reported)

Yeah I got the same hunch, just wanted to cross check with you.
 
> > Would it make sense for `parse_filename` to at least return `p.filename = filename` in some cases?
> 
> Not sure, the `[]` clearly indicates failed state, if the return will contain partially filled structure, the test for failed state will be more intricated. the function that calls `parse_filename` knows already the file name to parse

yeah that makes sense: parse_filename does enough already, so let's leave it at that

> > The reason I am wondering about it is also to because what should be the expected behavior of the following?
> 
> I'll say it should put in dataset maximum files that still follow bids syntax. Just makes it easier for unregulated bids datasets.

Will cross check how pybids handles those cases: I would not mind if the general behavior was fairly similar across tools.

**Comment by Remi-Gau** [comment]
>     * Adding `'tolerant', false` where it make sense

For now, I have removed the ones you had added that were in `create_filename`, `create_path` and `derivatives_json`

File: +bids/+internal/parse_filename.m [hunk 1]
```diff
@@ -78,39 +76,81 @@
...
+  for i = 1:numel(parts)
+
     try
-      [d, dummy] = regexp(parts{i}, '(?:\-)+', 'split', 'match'); %#ok<ASGLU>
-      p.entities.(d{1}) = d{2};
-    catch
-      msg = sprintf(['Entity-label pair %s of file %s is not valid.\n', ...
-                     'This could also be to a suffix with an extra _'], parts{i}, filename);
+
+      if isempty(parts{i})
```
Comment (Remi-Gau) [inline +bids/+internal/parse_filename.m:95]: am struggling to come up with a test case that hits this. Can you think of one?

**Comment by Remi-Gau** [comment]
> > > The reason I am wondering about it is also to because what should be the expected behavior of the following?
> > 
> > 
> > I'll say it should put in dataset maximum files that still follow bids syntax. Just makes it easier for unregulated bids datasets.
> 
> Will cross check how pybids handles those cases: I would not mind if the general behavior was fairly similar across tools.

Checked with my fellow bids maintainers and in brief pybids has an extra "validate" flag (lightweight filename based validation similar to what is done by parse_filename): if this is set to "false", then even bids invalid files will be indexed.

I think that implementing this behavior could be done but it is beyond the scope of this PR and I suspect it might be potentially annoying enough to implement: I suggest we wait till we have users requests for it.

File: +bids/+internal/parse_filename.m (hunk 1 above)
Comment (Remi-Gau) [reply +bids/+internal/parse_filename.m:115]: FYI I am ok to leave it in but I might add a comment to mention that this is unlikely to ever happen.

Comment (nbeliy) [reply +bids/+internal/parse_filename.m:115]: Just to cover weird cases with double '_':
```matlab
[parts, dummy] = regexp('_a_', '(?:_)+', 'split', 'match')

parts =

  1x3 cell array

    {0x0 char}    {'a'}    {0x0 char}
```

Comment (nbeliy) [reply +bids/+internal/parse_filename.m:115]: Actually, it just for cases of starting and ending with '_', it is nt catching duble underscore:
```matlab
parts, dummy] = regexp('a__b', '(?:_)+', 'split', 'match')

parts =

  1x2 cell array

    {'a'}    {'b'}


dummy =

  1x1 cell array

    {'__'}
```

Comment (Remi-Gau) [reply +bids/+internal/parse_filename.m:115]: Cool!!! 

Added a test to cover that error too.

**Comment by Remi-Gau** [comment]
Merging this now! Thanks @nbeliy

**Review by Remi-Gau — APPROVED** [review]

**Pull request merged**